### ISSUE TITLE
SuiteSparseQR_C.h: No such file or directory

### ISSUE THREAD

**Issue by nazariyv** [yig/PySPQR#11]
Cannot install your library. Installed SuiteSparse through conda like in the other issue. Did not work. In terms of how to install SuiteSparse through a link you provided in ReadME I do not know, there is no explanation either in your ReadME or their's web-page. Here is the error:

```
running install
running bdist_egg
running egg_info
writing sparseqr.egg-info/PKG-INFO
writing dependency_links to sparseqr.egg-info/dependency_links.txt
writing requirements to sparseqr.egg-info/requires.txt
writing top-level names to sparseqr.egg-info/top_level.txt
reading manifest file 'sparseqr.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'sparseqr.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
running build_ext
generating cffi module 'build/temp.linux-x86_64-3.6/sparseqr._sparseqr.c'
already up-to-date
building 'sparseqr._sparseqr' extension
gcc -pthread -B /home/nazariy/anaconda/envs/logging/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/include/suitesparse -I/home/nazariy/anaconda/envs/logging/include/python3.6m -c build/temp.linux-x86_64-3.6/sparseqr._sparseqr.c -o build/temp.linux-x86_64-3.6/build/temp.linux-x86_64-3.6/sparseqr._sparseqr.o
build/temp.linux-x86_64-3.6/sparseqr._sparseqr.c:492:29: fatal error: SuiteSparseQR_C.h: No such file or directory
compilation terminated.
error: command 'gcc' failed with exit status 1
```

**Comment by dswah** [comment]
i am seeing the same problem!

**Comment by dswah** [comment]
OK, i got it working 2 different ways. 
Note: i am running ubuntu 16.04 and py3.6

## 1
- `conda install -c conda-forge suitesparse`
- clone this repo
- locate the freshly installed `SuiteSparseQR_C.h` file (i used ubuntu's search...)
- change https://github.com/yig/PySPQR/blob/master/sparseqr/sparseqr_gen.py#L18 to point to the directory containing the file
- proceed with manual installation 

## 2
- install libsuitesparse (for apt-get linux) following http://ceres-solver.org/installation.html#linux
- `pip install sparseqr`

**Comment by yig** [comment]
@nazariyv What platform are you running on?

**Comment by yig** [comment]
@dswah I like the second version. Do you have a suggestion for the README?

**Comment by dswah** [comment]
@yig i guess the README already had the correct information, but it would have been good to clarify 2 things:
- debian/ubuntu linux wants `sudo apt-get install libsuitesparse-dev` 
- doing `conda install -c conda-forge suitesparse` isnt sufficient because it leaves the necessary files in the wrong place?

perhaps something can be done about the second point. 
is it possible for `PySPQR ` to look for suitesparse some other way?

**Comment by yig** [comment]
I pushed a change to the README.

**Comment by yig** [comment]
It may be possible to have PySPQR look for suitesparse in more locations. Feel free to look into it and propose a change. I don't have time now.

**Comment by nazariyv** [comment]
@yig same as you, ubuntu 16.04 and python 3.6

**Comment by sgbaird** [comment]
Would be nice to see an update on this. Still an issue several years later.

**Comment by sgbaird** [comment]
Maybe relevant? https://stackoverflow.com/questions/67440802/point-cmake-to-proper-python-header-inside-conda-env

**Comment by yig** [comment]
This may have been fixed. Can someone please check if it is still an issue?

**Comment by badalpardhi12** [comment]
> This may have been fixed. Can someone please check if it is still an issue?

Not yet, still having the issue installing it on windows VS Code Ananconda env.

**Comment by dallan-keylogic** [comment]
FWIW, I just experienced the same issue installing into Python 3.10.

**Comment by yig** [comment]
What is the path to your `SuiteSparseQR_C.h`? Are you using conda?

**Comment by dallan-keylogic** [comment]
It looks like I was running into two problems:

1) I was expecting pip to install `suitesparse`, but I had to manually do that through Condaforge. Might be good to put that explicitly as a step in the installation instructions instead of as a parenthetical comment in the dependencies.

2) I'm using Miniforge, which got installed into the `miniforge3` directory, while `sparseqr_gen.py` presently just looks for `miniforge`.

I now am getting a different error:

Edit: This error appears to be linked to C compilation, but I'm not sure how to fix it. [The solutions listed here](https://stackoverflow.com/questions/14372706/visual-studio-cant-build-due-to-rc-exe?newreg=f00ac10c2a4448ceb438f258e5c229a4) did not work for me.

```
In [2]: from sparseqr import qr
=== Wrapper module not compiled; compiling...
> c:\users\[REDACTED]\miniforge3\envs\idaes-fresh\lib\site-packages\sparseqr\sparseqr_gen.py(30)<module>()
-> if 'CONDA_DEFAULT_ENV' in os.environ:
(Pdb) c
generating C:\Users\[REDACTED]\miniforge3\envs\idaes-fresh\lib\site-packages\sparseqr\_sparseqr.c
(already up-to-date)
setting the current directory to 'C:\\Users\\[REDACTED]\\miniforge3\\envs\\idaes-fresh\\lib\\site-packages'
_sparseqr.c
LINK : fatal error LNK1181: cannot open input file 'spqr.lib'
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File ~\miniforge3\envs\idaes-fresh\lib\site-packages\sparseqr\sparseqr.py:15
     14 try:
---> 15     from ._sparseqr import ffi, lib
     16 except ImportError:

ModuleNotFoundError: No module named 'sparseqr._sparseqr'

During handling of the above exception, another exception occurred:

CalledProcessError                        Traceback (most recent call last)
File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\spawn.py:70, in spawn(cmd, search_path, verbose, dry_run, env)
     69 try:
---> 70     subprocess.check_call(cmd, env=_inject_macos_ver(env))
     71 except OSError as exc:

File ~\miniforge3\envs\idaes-fresh\lib\subprocess.py:369, in check_call(*popenargs, **kwargs)
    368         cmd = popenargs[0]
--> 369     raise CalledProcessError(retcode, cmd)
    370 return 0

CalledProcessError: Command '['C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.35.32215\\bin\\HostX86\\x64\\link.exe', '/nologo', '/INCREMENTAL:NO', '/LTCG', '/DLL', '/MANIFEST:EMBED,ID=2', '/MANIFESTUAC:NO', '/LIBPATH:C:\\Users\\[REDACTED]\\miniforge3\\envs\\idaes-fresh\\libs', '/LIBPATH:C:\\Users\\[REDACTED]\\miniforge3\\envs\\idaes-fresh', '/LIBPATH:C:\\Users\\[REDACTED]\\miniforge3\\envs\\idaes-fresh\\PCbuild\\amd64', '/LIBPATH:C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.35.32215\\lib\\x64', '/LIBPATH:C:\\Program Files (x86)\\Windows Kits\\10\\lib\\10.0.22000.0\\ucrt\\x64', '/LIBPATH:C:\\Program Files (x86)\\Windows Kits\\10\\\\lib\\10.0.22000.0\\\\um\\x64', 'spqr.lib', 'amd.lib', 'btf.lib', 'camd.lib', 'ccolamd.lib', 'cholmod.lib', 'colamd.lib', 'cxsparseklu.lib', 'lapack.lib', 'ldl.lib', 'lumfpack.lib', 'metis.lib', 'suitesparseconfig.lib', 'libblas.lib', '/EXPORT:PyInit__sparseqr', '.\\Release\\sparseqr\\_sparseqr.obj', '/OUT:.\\sparseqr\\_sparseqr.cp310-win_amd64.pyd', '/IMPLIB:.\\Release\\sparseqr\\_sparseqr.cp310-win_amd64.lib']' returned non-zero exit status 1181.

The above exception was the direct cause of the following exception:

DistutilsExecError                        Traceback (most recent call last)
File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\_msvccompiler.py:551, in MSVCCompiler.link(self, target_desc, objects, output_filename, output_dir, libraries, library_dirs, runtime_library_dirs, export_symbols, debug, extra_preargs, extra_postargs, build_temp, target_lang)
    550     log.debug('Executing "%s" %s', self.linker, ' '.join(ld_args))
--> 551     self.spawn([self.linker] + ld_args)
    552 except DistutilsExecError as msg:

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\_msvccompiler.py:560, in MSVCCompiler.spawn(self, cmd)
    559 with self._fallback_spawn(cmd, env) as fallback:
--> 560     return super().spawn(cmd, env=env)
    561 return fallback.value

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\ccompiler.py:1045, in CCompiler.spawn(self, cmd, **kwargs)
   1044 def spawn(self, cmd, **kwargs):
-> 1045     spawn(cmd, dry_run=self.dry_run, **kwargs)

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\spawn.py:76, in spawn(cmd, search_path, verbose, dry_run, env)
     75 except subprocess.CalledProcessError as err:
---> 76     raise DistutilsExecError(
     77         f"command {_debug(cmd)!r} failed with exit code {err.returncode}"
     78     ) from err

DistutilsExecError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.35.32215\\bin\\HostX86\\x64\\link.exe' failed with exit code 1181

During handling of the above exception, another exception occurred:

LinkError                                 Traceback (most recent call last)
File ~\miniforge3\envs\idaes-fresh\lib\site-packages\cffi\ffiplatform.py:48, in _build(tmpdir, ext, compiler_verbose, debug)
     47 set_verbosity(compiler_verbose)
---> 48 dist.run_command('build_ext')
     49 cmd_obj = dist.get_command_obj('build_ext')

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\dist.py:995, in Distribution.run_command(self, command)
    992 # Postpone defaults until all explicit configuration is considered
    993 # (setup() args, config files, command line and plugins)
--> 995 super().run_command(command)

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\dist.py:973, in Distribution.run_command(self, command)
    972 cmd_obj.ensure_finalized()
--> 973 cmd_obj.run()
    974 self.have_run[command] = True

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\command\build_ext.py:99, in build_ext.run(self)
     98 old_inplace, self.inplace = self.inplace, False
---> 99 _build_ext.run(self)
    100 self.inplace = old_inplace

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\command\build_ext.py:359, in build_ext.run(self)
    358 # Now actually compile and link everything.
--> 359 self.build_extensions()

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\command\build_ext.py:476, in build_ext.build_extensions(self)
    475 else:
--> 476     self._build_extensions_serial()

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\command\build_ext.py:502, in build_ext._build_extensions_serial(self)
    501 with self._filter_build_errors(ext):
--> 502     self.build_extension(ext)

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\command\build_ext.py:264, in build_ext.build_extension(self, ext)
    263     self.compiler = self.shlib_compiler
--> 264 _build_ext.build_extension(self, ext)
    265 if ext._needs_stub:

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\command\build_ext.py:581, in build_ext.build_extension(self, ext)
    579 language = ext.language or self.compiler.detect_language(sources)
--> 581 self.compiler.link_shared_object(
    582     objects,
    583     ext_path,
    584     libraries=self.get_libraries(ext),
    585     library_dirs=ext.library_dirs,
    586     runtime_library_dirs=ext.runtime_library_dirs,
    587     extra_postargs=extra_args,
    588     export_symbols=self.get_export_symbols(ext),
    589     debug=self.debug,
    590     build_temp=self.build_temp,
    591     target_lang=language,
    592 )

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\ccompiler.py:758, in CCompiler.link_shared_object(self, objects, output_filename, output_dir, libraries, library_dirs, runtime_library_dirs, export_symbols, debug, extra_preargs, extra_postargs, build_temp, target_lang)
    743 def link_shared_object(
    744     self,
    745     objects,
   (...)
    756     target_lang=None,
    757 ):
--> 758     self.link(
    759         CCompiler.SHARED_OBJECT,
    760         objects,
    761         output_filename,
    762         output_dir,
    763         libraries,
    764         library_dirs,
    765         runtime_library_dirs,
    766         export_symbols,
    767         debug,
    768         extra_preargs,
    769         extra_postargs,
    770         build_temp,
    771         target_lang,
    772     )

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\setuptools\_distutils\_msvccompiler.py:553, in MSVCCompiler.link(self, target_desc, objects, output_filename, output_dir, libraries, library_dirs, runtime_library_dirs, export_symbols, debug, extra_preargs, extra_postargs, build_temp, target_lang)
    552     except DistutilsExecError as msg:
--> 553         raise LinkError(msg)
    554 else:

LinkError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.35.32215\\bin\\HostX86\\x64\\link.exe' failed with exit code 1181

During handling of the above exception, another exception occurred:

VerificationError                         Traceback (most recent call last)
Cell In[2], line 1
----> 1 from sparseqr import qr

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\sparseqr\__init__.py:23
     20 __version__ = '1.3'
     22 # import the important things into the package's top-level namespace.
---> 23 from .sparseqr import qr, rz, solve, permutation_vector_to_matrix, qr_factorize,qmult

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\sparseqr\sparseqr.py:19
     17 print( "=== Wrapper module not compiled; compiling..." )
     18 from .sparseqr_gen import main
---> 19 main()
     20 print( "=== ...compiled." )
     22 from ._sparseqr import ffi, lib

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\sparseqr\sparseqr_gen.py:488, in main()
    485 def main():
    486     ## Two dirnames because ffibuilder.set_source()
    487     ## passes the module name: "sparseqr._sparseqr"
--> 488     ffibuilder.compile( verbose = True, tmpdir = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) )

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\cffi\api.py:727, in FFI.compile(self, tmpdir, verbose, target, debug)
    725     raise ValueError("set_source() must be called before compile()")
    726 module_name, source, source_extension, kwds = self._assigned_source
--> 727 return recompile(self, module_name, source, tmpdir=tmpdir,
    728                  target=target, source_extension=source_extension,
    729                  compiler_verbose=verbose, debug=debug, **kwds)

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\cffi\recompiler.py:1581, in recompile(ffi, module_name, preamble, tmpdir, call_c_compiler, c_file, source_extension, extradir, compiler_verbose, target, debug, uses_ffiplatform, **kwds)
   1579         print('%s %r' % (msg, os.path.abspath(tmpdir)))
   1580     os.chdir(tmpdir)
-> 1581     outputfilename = ffiplatform.compile('.', ext,
   1582                                          compiler_verbose, debug)
   1583 finally:
   1584     os.chdir(cwd)

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\cffi\ffiplatform.py:20, in compile(tmpdir, ext, compiler_verbose, debug)
     18 saved_environ = os.environ.copy()
     19 try:
---> 20     outputfilename = _build(tmpdir, ext, compiler_verbose, debug)
     21     outputfilename = os.path.abspath(outputfilename)
     22 finally:
     23     # workaround for a distutils bugs where some env vars can
     24     # become longer and longer every time it is used

File ~\miniforge3\envs\idaes-fresh\lib\site-packages\cffi\ffiplatform.py:54, in _build(tmpdir, ext, compiler_verbose, debug)
     52         set_threshold(old_level)
     53 except (CompileError, LinkError) as e:
---> 54     raise VerificationError('%s: %s' % (e.__class__.__name__, e))
     55 #
     56 return soname

VerificationError: LinkError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.35.32215\\bin\\HostX86\\x64\\link.exe' failed with exit code 118
```

**Comment by yig** [comment]
1. I updated the README. Hopefully it's clearer now.
2. `sparseqr_gen.py` has `miniconda3` in its list of directories to search: <https://github.com/yig/PySPQR/blob/c8fdaeaadc72e3e878670d016df774a8675c1907/sparseqr/sparseqr_gen.py#L33>. I'm not sure how to debug this. If you find a solution, I can incorporate it.

**Comment by dallan-keylogic** [comment]
Hmm. My local version did not have it included. Probably the version on pip is out of date.

```
(idaes-fresh) C:\Users\[REDACTED]\Repos\mea-flowsheet>pip show sparseqr
Name: sparseqr
Version: 1.3
Summary: Python wrapper for SuiteSparseQR
Home-page:
Author: Yotam Gingold
Author-email: yotam@yotamgingold.com
License: Public Domain CC0
Location: c:\users\[REDACTED]\miniforge3\envs\idaes-fresh\lib\site-packages
Requires: cffi, numpy, scipy, setuptools
Required-by:
````

Thank you for replying.

**Comment by yig** [comment]
That line is there on the version on PyPI. It's in version 1.3:

<https://github.com/yig/PySPQR/blob/v1.3/sparseqr/sparseqr_gen.py#L33>

**Comment by dallan-keylogic** [comment]
No idea what's going on. Maybe I got it through Conda instead? I definitely went into the file, added a pdb trace, looked at the directories that it iterated over, and noticed that `miniforge3` was missing.

**Comment by yig** [comment]
I have changed the way that the package installs itself and finds the suite sparse dependency. Hopefully that fixed this issue. I bumped the version to 1.4. Please try again.
